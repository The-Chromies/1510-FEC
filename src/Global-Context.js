/* eslint-disable react/destructuring-assignment */
/* eslint-disable react/prop-types */
/* eslint-disable prefer-const */
import React, { useState, createContext } from 'react';
import axios from 'axios';
import uuid from 'node-uuid';
import LocalServer from './env/localServ';

export const ContactContext = createContext();

export const ContactContextProvider = (props) => {
  // global state accessible to all components
  let [productId, setProductId] = useState(18085);
  const [revCount, setRevCount] = useState(0);
  const [avgRating, setAvgRating] = useState(0);
  // arr of products added to outfit list in rc
  const [outfitProduct, setOutfitProduct] = useState([]);
  // arr of styles added to outfit list in rc
  const [outfitStyle, setOutfitStyle] = useState([]);
  // selected product
  const [product, setProduct] = useState(null);
  // selected styles obj
  const [styles, setStyles] = useState(null);
  const localServer = LocalServer;

  // click tracking function
  const clickTracker = (widgetName, e) => {
  // post to /interactions endpoint w/ element's class, widget, & time of click
    let date = new Date();
    let clickData = {
      element: e.target.className,
      widget: widgetName,
      time: date.toTimeString(),
    };
    axios.post(`http://${localServer}:3000/interactions`, clickData)
      .then((res) => {
        // console.log('SUCCESSFUL POST TO INTERACTIONS');
        console.log(res);
      })
      .catch((err) => {
        // console.log('FAILED TO POST TO INTERACTIONS');
        console.log(err);
      });
  };

  const generateStarImage = (ratingValue, keyId) => {
    const starCount = ratingValue > 0 ? (Math.round((ratingValue) * 4) / 4).toFixed(2) : 0;

    let remainder = 0;
    remainder = starCount - remainder;
    const starArr = [];
    for (let j = 0; j < 5; j += 1) {
      switch (remainder) { // 4.25
        case 0.25:
          starArr.push(<div key={uuid()}><img src="public/icons/star/quarterStar.png" alt="" className="star-image" /></div>);
          remainder -= 0.25;
          break;
        case 0.50:
          starArr.push(<div key={uuid()}><img src="public/icons/star/halfStar.png" alt="" className="star-image" /></div>);
          remainder -= 0.50;
          break;
        case 0.75:
          starArr.push(<div key={uuid()}><img src="public/icons/star/threeQuarterStar.png" alt="" className="star-image" /></div>);
          remainder -= 0.75;
          break;
        case 1:
          starArr.push(<div key={uuid()}><img src="public/icons/star/fullStar.png" alt="" className="star-image" /></div>);
          remainder -= 1;
          break;
        case 0:
          starArr.push(<div key={uuid()}><img src="public/icons/star/emptyStar.png" alt="" className="star-image" /></div>);
          break;
        default:
          starArr.push(<div key={uuid()}><img src="public/icons/star/fullStar.png" alt="" className="star-image" /></div>);
          remainder -= 1;
          break;
      }
    }
    return (<div className="inline-star" style={{ display: 'inline-flex' }} key={uuid()}>{starArr}</div>);
  };

  // get selected product & style for ov & rc components
  const getProduct = (id) => {
    axios.get(`http://${localServer}:3000/related/product/${id}`)
      .then((results) => {
        console.log('OUTFITPRODUCT!!', results.data);
        let outfitProductList = [...outfitProduct];
        outfitProductList.push(results.data);
        setOutfitProduct(outfitProductList);
      })
      .catch((err) => {
        console.log('err getting outfit product info:', err);
      });
  };

  const getStyle = (id) => {
    axios.get(`http://${localServer}:3000/overview/styles/${id}`)
      .then((results) => {
        let outfitStyleList = [...outfitStyle];
        outfitStyleList.push(results.data);
        setOutfitStyle(outfitStyleList);
      });
  };

  // add product to outfit list
  const handelAddOutfit = (id) => {
    setProductId(id);
    getProduct(id);
    getStyle(id);
  };

  return (
    <ContactContext.Provider value={{
      productId,
      setProductId,
      generateStarImage,
      revCount,
      setRevCount,
      avgRating,
      product,
      styles,
      setAvgRating,
      clickTracker,
      outfitProduct,
      setOutfitProduct,
      outfitStyle,
      setOutfitStyle,
      getProduct,
      getStyle,
      handelAddOutfit,
      localServer,
    }}
    >
      {props.children}
    </ContactContext.Provider>
  );
};
