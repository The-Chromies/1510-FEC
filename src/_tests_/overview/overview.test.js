/* eslint-disable max-len */
/* eslint-disable no-undef */
import React from 'react';
import Enzyme, { shallow, mount, render } from 'enzyme';
import renderer from 'react-test-renderer';
import axios from 'axios';
import { ContactContextProvider } from '../../Global-Context';
import { product, style } from '../testData';
import Overview from '../../components/overview/ovMain';
import ImageGallery from '../../components/overview/components/imageGallery';
import ProductInfo from '../../components/overview/components/productInfo';
import StyleSelector from '../../components/overview/components/styleSelector';
import AddToCart from '../../components/overview/components/addToCart';

it('matches the overview snapshot', () => {
  const snapshot = renderer
    .create(<ContactContextProvider><Overview /></ContactContextProvider>)
    .toJSON();
  expect(snapshot).toMatchSnapshot();
});

// let testReviewList = [];
// if (props.testOn) {
//   productId = props.productId;
//   testReviewList = props.reviewList;
// }
const imgProps = {
  selected: style,
  currentIndex: 1,
  generateStarImage: () => {},
  resetIndex: () => {},
  sendClick: () => {},
};

it('matches the image gallery snapshot', () => {
  const snapshot = renderer
    .create(<ContactContextProvider><ImageGallery {...imgProps} /></ContactContextProvider>)
    .toJSON();
  expect(snapshot).toMatchSnapshot();
});

// npm test -- -u to update a snapshot

jest.mock('axios');

const getInfo = () => Promise.all([
  (axios.get.mockResolvedValue({ data: product }))(),
  (axios.get.mockResolvedValue({ data: style }))(),
]);

describe('Overview', () => {
  let wrapper;

  it('should render without props', () => {
    wrapper = shallow(<ContactContextProvider><Overview getInfo={getInfo} /></ContactContextProvider>);
    // expect(wrapper.exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should have image gallery', () => {
    wrapper = shallow(<ContactContextProvider><Overview getInfo={getInfo} /></ContactContextProvider>);
    expect(wrapper.find(<ImageGallery />)).toBeTruthy();
  });

  it('should have product info', () => {
    expect(wrapper.find(<ProductInfo styles={style} />)).toBeTruthy();
  });

  it('should have style selector', () => {
    expect(wrapper.find(<StyleSelector />)).toBeTruthy();
  });

  it('should have add to cart', () => {
    expect(wrapper.find(<AddToCart selected={{}} />)).toBeTruthy();
  });

  // it('should conditionally render child components', () => {
  //   wrapper = shallow(<ContactContextProvider><Overview /></ContactContextProvider>);
  //   expect(wrapper.contains(<ImageGallery />)).toBeTruthy();
  // });
});
