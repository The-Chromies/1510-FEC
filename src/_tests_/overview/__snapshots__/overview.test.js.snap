// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overview should render without props 1`] = `ShallowWrapper {}`;

exports[`matches the overview snapshot 1`] = `
<div
  className="container"
>
  <div
    className="overview-container row"
  >
    <div
      className="col-lg-8 col-md-6 col-12"
      s={12}
    />
    <div
      className="col-lg-4 col-md-6 col-12"
      s={12}
    />
  </div>
</div>
`;
